File: backend/session_manager.py
import time
from typing import Dict, List, Any

SESSION_EXPIRY_SECONDS = 600  # 10 minutes

class SessionManager:
    def __init__(self):
        self.sessions: Dict[str, Dict[str, Any]] = {}

    def get_session(self, session_id: str) -> List[Dict[str, str]]:
        session = self.sessions.get(session_id)
        now = time.time()
        if session and (now - session["last_access"] < SESSION_EXPIRY_SECONDS):
            session["last_access"] = now
            return session["messages"]
        # New session or expired
        self.sessions[session_id] = {
            "last_access": now,
            "messages": []
        }
        return []

    def update_session(self, session_id: str, messages: List[Dict[str, str]]) -> None:
        self.sessions[session_id] = {
            "last_access": time.time(),
            "messages": messages
        }

File: utils/sessionManager.js
const sessions = new Map();

// Cleanup expired sessions every minute
setInterval(() => {
    const now = Date.now();
    for (const [sessionId, session] of sessions) {
        if (now - session.lastAccessed > 30 * 60 * 1000) { // 30 minutes
            sessions.delete(sessionId);
        }
    }
}, 60 * 1000);

export const sessionManager = {
    createSession() {
        const sessionId = generateId();
        const session = {
            id: sessionId,
            state: {
                operation: null,
                extractedFields: {},
                confirmed: false,
                query: null
            },
            history: [],
            createdAt: Date.now(),
            lastAccessed: Date.now()
        };
        sessions.set(sessionId, session);
        return session;
    },

    getSession(sessionId) {
        const session = sessions.get(sessionId);
        if (session) {
            session.lastAccessed = Date.now();
        }
        return session;
    },

    updateSession(sessionId, updates) {
        const session = this.getSession(sessionId);
        if (!session) return null;

        Object.assign(session.state, updates);
        session.lastAccessed = Date.now();
        return session;
    },

    addHistory(sessionId, entry) {
        const session = this.getSession(sessionId);
        if (session) {
            session.history.push({
                role: entry.role,
                content: entry.content,
                timestamp: new Date()
            });
        }
    }
};

function generateId() {
    return Math.random().toString(36).substring(2, 15) +
        Math.random().toString(36).substring(2, 15);
}